### PR TITLE
fix `useWalletClient` and `useConnectorClient` performance issues

### PR DESCRIPTION
**Pull request by BrickheadJohnny**
## Description

Fixes the performance issues discussed in #3738 
Thank you for all the help @Yuripetusko!

## Additional Information

Before submitting this issue, please make sure you do the following.

- [x] Read the [contributing guide](https://wagmi.sh/dev/contributing)
- [x] Added documentation related to the changes made.
- [x] Added or updated tests (and snapshots) related to the changes made.

### CODE REVIEW

File: packages/react/src/hooks/useConnectorClient.ts [hunk 1]
```diff
@@ -80,11 +80,22 @@ export function useConnectorClient<
...
   const enabled = Boolean(status !== 'disconnected' && (query.enabled ?? true))
 
+  const addressRef = useRef<UseAccountReturnType<config>['address']>(address)
+
   // biome-ignore lint/nursery/useExhaustiveDependencies: `queryKey` not required
   useEffect(() => {
-    // invalidate when address changes
-    if (address) queryClient.invalidateQueries({ queryKey })
-    else queryClient.removeQueries({ queryKey }) // remove when account is disconnected
+    if (!address) {
```
Comment (Yuripetusko) [inline packages/react/src/hooks/useConnectorClient.ts:87]: wouldn't this always removeQueries on page refresh? Not sure if there are edge cases like offline-first apps etc. Although perhaps re-initialising new walletClient on page refresh is a good thing

Comment (BrickheadJohnny) [reply packages/react/src/hooks/useConnectorClient.ts:87]: Yup, good point & I think it is actually okay to revalidate on refresh

Comment (Yuripetusko) [reply packages/react/src/hooks/useConnectorClient.ts:87]: Yeah I suppose you are right, client is not that critical to re-create on page load and forget the old one. As oppose to things like account, that is more important to get from persistent storage for offline-first apps

Comment (jxom) [reply packages/react/src/hooks/useConnectorClient.ts:87]: idk if we should be removing queries on refresh if we don't need to

Comment (Yuripetusko) [reply packages/react/src/hooks/useConnectorClient.ts:87]: ok yea, in this case @BrickheadJohnny perhaps you can check for `!!addressRef.current` ? and I would probably remove it from useRef initialisation, this feels like can lead to weird behaviour. I think it's better to have it null or undefined by default. wdyt?

File: packages/react/src/hooks/useConnectorClient.ts [hunk 2]
```diff
@@ -80,11 +80,22 @@ export function useConnectorClient<
   })
   const enabled = Boolean(status !== 'disconnected' && (query.enabled ?? true))
 
+  const addressRef = useRef<UseAccountReturnType<config>['address']>(address)
```
Comment (Yuripetusko) [inline packages/react/src/hooks/useConnectorClient.ts:83]:
```suggestion
  const addressRef = useRef<UseAccountReturnType<config>['address']>()
```

File: packages/react/src/hooks/useConnectorClient.ts (hunk 1 above)
Comment (Yuripetusko) [inline packages/react/src/hooks/useConnectorClient.ts:87]:
```suggestion
    if (!address && !!addressRef.current) {
```

Comment (Yuripetusko) [reply packages/react/src/hooks/useConnectorClient.ts:87]: Actually I take it back

**Comment by BrickheadJohnny** [comment]
Hey @tmm, according to GitHub, you are the code owner of the `useWalletClient.test.ts` & `useConnectorClient.test.ts` files, could I ask for some guidance here please? 👀 
I'd like to add tests for this bugfix, but wonder how could we re-create this case which I described in the original issue description (or basically the one which you could try in the linked Codesandbox) with the `renderHook` function? 🤔 So the test should look something like this:

1. render the `useWalletClient` hook
2. render a child component, which also calls this hook
3. WalletClient's `uid` shouldn't change after the child (re)renders

If I just call the `renderHook` function 2 times, it'll wrap the 2 hooks in 2 different `WagmiProvider`s, so that wouldn't really work. Should we create & mount actual components in tests with the [render](https://testing-library.com/docs/react-testing-library/api#render) method? (I haven't tried that solution yet, because after all wagmi is a hooks library, and I see that you don't do that in other tests)

**Comment by BrickheadJohnny** [comment]
> Hey @tmm, according to GitHub, you are the code owner of the `useWalletClient.test.ts` & `useConnectorClient.test.ts` files, could I ask for some guidance here please? 👀 I'd like to add tests for this bugfix, but wonder how could we re-create this case which I described in the original issue description (or basically the one which you could try in the linked Codesandbox) with the `renderHook` function? 🤔 So the test should look something like this:
> 
> 1. render the `useWalletClient` hook
> 2. render a child component, which also calls this hook
> 3. WalletClient's `uid` shouldn't change after the child (re)renders with a different `key`
> 
> If I just call the `renderHook` function 2 times, it'll wrap the 2 hooks in 2 different `WagmiProvider`s, so that wouldn't really work. Should we create & mount actual components in tests with the [render](https://testing-library.com/docs/react-testing-library/api#render) method? (I haven't tried that solution yet, because wagmi is a hooks library after all, and I see that you don't do that in other tests)

I ended up rendering components & basically re-created the behavior which I mentioned in the original issue. I'm pretty happy with it (except that `wait(200)`), but please let me know if you think there's a simper solution for this ;)

I also fixed the `useWalletClient` tests, so now they actually test the proper hook (we talked about those on Discord yesterday)

File: packages/react/tsconfig.json [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {
   "extends": "./tsconfig.build.json",
-  "include": ["src/**/*.ts", "test/**/*.ts"],
+  "include": ["src/**/*.ts?(x)", "test/**/*.ts?(x)"],
```
Comment (Yuripetusko) [inline packages/react/tsconfig.json:3]: Was testing this, and this pattenr matching wasn't working well for me (had ts error), but this worked:

```
"include": ["src/**/*.ts","src/**/*.tsx", "test/**/*.ts", "test/**/*.tsx"],
  "compilerOptions": {
    "jsx": "preserve"
  },
```

**Comment by Yuripetusko** [comment]
Just ran these tests with and without your fix, and indeed, without the fix, the uuid of wallet client in re-rendered component differs, but with fix, it's the same. 👍

File: packages/react/tsconfig.json (hunk 3 above)
Comment (BrickheadJohnny) [reply packages/react/tsconfig.json]: Hmm, interesting, changed it in 88ec9072e3b7b74d6b84aeecc3e866d7296db239 to make sure it works for everyone, thanks!

File: .changeset/olive-ravens-eat.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"wagmi": patch
+---
+
+fix `useWalletClient` and `useConnectorClient` performance issues
```
Comment (tmm) [inline .changeset/olive-ravens-eat.md:5]:
```suggestion
Removed unnecessary re-renders from  `useConnectorClient` and `useWalletClient`.
```

**Pull request merged**